docs/proposals/20200615-external-remediation.md

```diff
@@ -0,0 +1,299 @@
+#### Story 2
+As an admin of a hardware based cluster, I would like unhealthy nodes to be power-cycled, so that I can detect hardware issues faster.
+#### Story 3
+As an admin of a hardware based cluster, I would like the system to keep attempting to power-cycle unhealthy nodes, so that they are automatically added back to the cluster when I fix the underlying problem.
+#### Story 4
+As an admin of a large bare metal cluster, I want to auto remediate nodes but do so grouped by failure domain (i.e. by rack). All nodes deemed unhealthy in a given rack should be cordoned first before each are drained.
+
+### Implementation Details/Notes/Constraints
+
+#### Flow
+The default MHC remediation strategy continues to be an condition-based flow where the owning controller deletes the Machine (and by inference the Node) objects after performing any relevant cleanup.
+
+In order to enable a different remediation strategy for a set of Machines, the admin must provide a reference to a provider-specific template object as part of the MachineHealthCheck CR.
+
+When a Machine enters an unhealthy state, the MHC will:
+* Look up the referenced template
+* Instantiate the template (for simplicity, we will refer to this as a External Machine Remediation CR, or EMR)
+* Force the name and namespace to match the unhealthy Machine
+* Save the new object in etcd
+
+We use the same name and namespace for the External Machine Remediation CR to ensure uniqueness and lessen the possibility for multiple parallel remediations of the same Machine. 
+
+The lifespan of the EMRs is that of the remediation process and they are not intended to be a record of past events.  The EMR will also contain an ownerRef to the Machine, to ensure that it does not outlive the Machine it references.
+
+The only signaling between the MHC and the external controller watching EMR CRs is the creation and deletion of the EMR itself.  Any actions or changes that admins should be informed about should be emitted as events for consoles and UIs to consume if necessary.  They are informational only and do not result in or expect any behaviour from the MHC, Node, or Machine as a result.
+
+When the external remediation controller detects the new EMR it starts remediation and performs whatever actions it deems appropriate until the EMR is deleted by the MHC. It is a detail of the ERC when and how to retry remediation in the event that a EMR is not deleted after the ERC considers remediation complete. 
+
+The ERC may wish to register a finalizer on its CR to ensure it has an opportunity to perform any additional cleanup in the case that the unhealthy state was transient and the Node returned to a healthy state prior to the completion of the full custom ERC flow.
+
+#### Proposed CRD changes and additions
+
+    type MachineHealthCheckSpec struct {
+        ...
+    
+        // +optional
+        RemediationTemplate ObjectReference `json:"remediationTemplate,omitempty"`
```

Comment:
nit
go doc missing (It will help to clarify the intent of this doc)
Also, what about adding a prefix like "external" or "custom" in order to make it more explicit that this field is optional and to suggest that if the value is empty a default behavior applies?

Reply:
"ExternalRemediationTemplate" could make sense.

Reply:
This should also likely be a custom type scoped to the exact type of information needed, rather than corev1.ObjectReference.

Thinking aloud, I suspect the following information would be needed:
apigroup
kind
name

We should also likely avoid the use of storing the version for the referenced type, otherwise we potentially need to worry about migration strategies for these loosely coupled types.

However if we don't encode the version, then instantiating the template would require a way to lookup (through discovery or some other means) the correct version of the resource to create.

Reply:
> "ExternalRemediationTemplate" could make sense.

:+1: 
> This will need to be a pointer

:+1: 

> I suspect the following information would be needed:
apigroup
kind
name

isn't it what we have in MHC CR under `externalRemediationTemplate`?